File: codoxear/git_context.py
"""Resolve git branch and GitHub PR context for a session's working directory.

Used by the session-list endpoint and the per-session repo detail endpoint to
surface branch and PR badges in the workspace UI without blocking session list
polling on subprocess calls.
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import threading
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

# Per-call timeouts. Kept short so a slow git or network call cannot block
# session-list polling. Branch lookup is local-only and very fast; PR lookup
# may hit the network through `gh`.
BRANCH_TIMEOUT_S = float(os.environ.get("CODEX_WEB_BRANCH_TIMEOUT_S", "2.0"))
PR_TIMEOUT_S = float(os.environ.get("CODEX_WEB_PR_TIMEOUT_S", "4.0"))

# Cache TTLs.
BRANCH_TTL_S = float(os.environ.get("CODEX_WEB_BRANCH_TTL_S", "30.0"))
PR_TTL_S = float(os.environ.get("CODEX_WEB_PR_TTL_S", "120.0"))

# How long to trust a `gh auth status` result.
GH_AUTH_TTL_S = float(os.environ.get("CODEX_WEB_GH_AUTH_TTL_S", "300.0"))


Availability = Literal["ok", "no-gh", "no-pr", "not-a-repo", "error"]


@dataclass(frozen=True)
class PullRequest:
    number: int
    title: str
    state: str  # OPEN | DRAFT | CLOSED | MERGED
    url: str
    is_draft: bool
    head_ref_name: str

    def to_summary(self) -> dict[str, Any]:
        """Compact form for the session-list payload."""
        state = "DRAFT" if self.is_draft and self.state == "OPEN" else self.state
        return {"number": int(self.number), "state": state}

    def to_dict(self) -> dict[str, Any]:
        return {
            "number": int(self.number),
            "title": self.title,
            "state": self.state,
            "url": self.url,
            "is_draft": bool(self.is_draft),
            "head_ref_name": self.head_ref_name,
        }


@dataclass(frozen=True)
class RepoContext:
    git_branch: str | None
    pr: PullRequest | None
    availability: Availability
    cwd: str = ""

    def to_detail_dict(self) -> dict[str, Any]:
        return {
            "cwd": self.cwd,
            "git_branch": self.git_branch,
            "pr": self.pr.to_dict() if self.pr else None,
            "availability": self.availability,
        }

    def pr_summary(self) -> dict[str, Any] | None:
        return self.pr.to_summary() if self.pr else None


@dataclass
class _CacheEntry:
    context: RepoContext
    expires_at: float


# Per-cwd cache + lock dictionaries. Locks ensure only one resolver runs per
# cwd at a time so concurrent pollers do not stampede `git`/`gh`.
_cache: dict[str, _CacheEntry] = {}
_cache_lock = threading.Lock()
_cwd_locks: dict[str, threading.Lock] = {}

# Global gh-auth cache state.
_gh_auth_state: dict[str, Any] = {"ok": False, "checked_at": 0.0}
_gh_auth_lock = threading.Lock()


def _now() -> float:
    return time.monotonic()


def _cwd_lock_for(key: str) -> threading.Lock:
    with _cache_lock:
        lock = _cwd_locks.get(key)
        if lock is None:
            lock = threading.Lock()
            _cwd_locks[key] = lock
        return lock


def _resolve_cwd(cwd: Path) -> Path | None:
    try:
        return Path(cwd).expanduser().resolve()
    except (OSError, RuntimeError):
        return None


def _run(
    cmd: list[str],
    *,
    cwd: Path,
    timeout_s: float,
) -> tuple[int, str, str]:
    """Run a subprocess and return (returncode, stdout, stderr).

    Returns (-1, "", "<reason>") on FileNotFoundError or TimeoutExpired so the
    caller can branch on a single tuple shape.
    """
    try:
        proc = subprocess.run(
            cmd,
            cwd=str(cwd),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=timeout_s,
            check=False,
        )
    except FileNotFoundError as exc:
        return -1, "", f"missing executable: {exc}"
    except subprocess.TimeoutExpired:
        return -1, "", "timeout"
    return (
        proc.returncode,
        proc.stdout.decode("utf-8", errors="replace"),
        proc.stderr.decode("utf-8", errors="replace"),
    )


def _resolve_branch(cwd: Path) -> tuple[str | None, bool]:
    """Return (branch_or_short_sha, is_repo).

    On detached HEAD returns the short SHA so the UI can still render
    something meaningful. Returns (None, False) when cwd is not a git tree.
    """
    code, out, _err = _run(
        ["git", "rev-parse", "--is-inside-work-tree"],
        cwd=cwd,
        timeout_s=BRANCH_TIMEOUT_S,
    )
    if code != 0 or out.strip().lower() != "true":
        return None, False

    code, out, _err = _run(
        ["git", "symbolic-ref", "--quiet", "--short", "HEAD"],
        cwd=cwd,
        timeout_s=BRANCH_TIMEOUT_S,
    )
    if code == 0:
        branch = out.strip()
        if branch:
            return branch, True

    # Detached HEAD: fall back to short SHA.
    code, out, _err = _run(
        ["git", "rev-parse", "--short", "HEAD"],
        cwd=cwd,
        timeout_s=BRANCH_TIMEOUT_S,
    )
    if code == 0:
        sha = out.strip()
        return (sha or None), True
    return None, True


def _gh_available() -> bool:
    return shutil.which("gh") is not None


def _check_gh_auth(*, refresh: bool = False) -> bool:
    """Cached `gh auth status` check. False when gh is missing or unauthenticated."""
    if not _gh_available():
        with _gh_auth_lock:
            _gh_auth_state["ok"] = False
            _gh_auth_state["checked_at"] = _now()
        return False
    with _gh_auth_lock:
        last = float(_gh_auth_state.get("checked_at") or 0.0)
        if not refresh and (_now() - last) < GH_AUTH_TTL_S:
            return bool(_gh_auth_state.get("ok"))
    code, _out, _err = _run(
        ["gh", "auth", "status"],
        cwd=Path.cwd(),
        timeout_s=PR_TIMEOUT_S,
    )
    ok = code == 0
    with _gh_auth_lock:
        _gh_auth_state["ok"] = ok
        _gh_auth_state["checked_at"] = _now()
    return ok


def _resolve_pr(cwd: Path) -> tuple[PullRequest | None, Availability]:
    """Look up PR metadata for the current branch via the gh CLI."""
    if not _gh_available():
        return None, "no-gh"
    if not _check_gh_auth():
        return None, "no-gh"
    code, out, err = _run(
        [
            "gh",
            "pr",
            "view",
            "--json",
            "number,title,state,url,isDraft,headRefName",
        ],
        cwd=cwd,
        timeout_s=PR_TIMEOUT_S,
    )
    if code == -1:
        return None, "error"
    if code != 0:
        # `gh pr view` exits non-zero when the branch has no associated PR.
        msg = (err or out).lower()
        if "no pull requests found" in msg or "no pr" in msg or "could not find" in msg:
            return None, "no-pr"
        return None, "error"
    try:
        data = json.loads(out)
    except ValueError:
        return None, "error"
    if not isinstance(data, dict):
        return None, "error"
    try:
        pr = PullRequest(
            number=int(data.get("number") or 0),
            title=str(data.get("title") or ""),
            state=str(data.get("state") or "").upper(),
            url=str(data.get("url") or ""),
            is_draft=bool(data.get("isDraft") or False),
            head_ref_name=str(data.get("headRefName") or ""),
        )
    except (TypeError, ValueError):
        return None, "error"
    if pr.number <= 0:
        return None, "no-pr"
    return pr, "ok"


def _compute(cwd: Path) -> RepoContext:
    branch, is_repo = _resolve_branch(cwd)
    if not is_repo:
        return RepoContext(
            git_branch=None, pr=None, availability="not-a-repo", cwd=str(cwd)
        )
    pr, availability = _resolve_pr(cwd)
    return RepoContext(
        git_branch=branch,
        pr=pr,
        availability=availability,
        cwd=str(cwd),
    )


def resolve_repo_context(cwd: Path | str, *, refresh: bool = False) -> RepoContext:
    """Resolve repo context for cwd, using a per-cwd TTL cache.

    Set ``refresh=True`` to bypass the cache and re-run `git`/`gh`. Branch
    lookups respect ``BRANCH_TTL_S`` and PR lookups respect ``PR_TTL_S``;
    the longer of the two is used for the combined entry.
    """
    resolved = _resolve_cwd(Path(cwd))
    if resolved is None or not resolved.exists() or not resolved.is_dir():
        return RepoContext(
            git_branch=None,
            pr=None,
            availability="not-a-repo",
            cwd=str(cwd),
        )
    key = str(resolved)
    if not refresh:
        with _cache_lock:
            entry = _cache.get(key)
            if entry is not None and entry.expires_at > _now():
                return entry.context
    lock = _cwd_lock_for(key)
    with lock:
        # Double-check after acquiring the per-cwd lock so concurrent waiters
        # share the freshly-computed result.
        if not refresh:
            with _cache_lock:
                entry = _cache.get(key)
                if entry is not None and entry.expires_at > _now():
                    return entry.context
        ctx = _compute(resolved)
        # Cache TTL: PR TTL when PR data was fetched, branch TTL otherwise.
        ttl = PR_TTL_S if ctx.availability in {"ok", "no-pr"} else BRANCH_TTL_S
        with _cache_lock:
            _cache[key] = _CacheEntry(context=ctx, expires_at=_now() + ttl)
        return ctx


def current_git_branch(cwd: Path) -> str | None:
    """Backwards-compatible thin wrapper used by ``server._current_git_branch``."""
    return resolve_repo_context(cwd).git_branch


def reset_caches_for_tests() -> None:
    """Clear all caches; intended for unit tests only."""
    with _cache_lock:
        _cache.clear()
        _cwd_locks.clear()
    with _gh_auth_lock:
        _gh_auth_state["ok"] = False
        _gh_auth_state["checked_at"] = 0.0


__all__ = [
    "Availability",
    "PullRequest",
    "RepoContext",
    "resolve_repo_context",
    "current_git_branch",
    "reset_caches_for_tests",
]

File: backend-rs/src/git_context.rs
use serde::Serialize;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::{Mutex, OnceLock};
use std::time::{Duration, Instant};

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct PullRequest {
    pub number: i64,
    pub title: String,
    pub state: String,
    pub url: String,
    pub is_draft: bool,
    pub head_ref_name: String,
}

impl PullRequest {
    pub fn to_summary(&self) -> Value {
        let state = if self.is_draft && self.state == "OPEN" {
            "DRAFT"
        } else {
            self.state.as_str()
        };
        json!({"number": self.number, "state": state})
    }

    pub fn to_dict(&self) -> Value {
        json!({
            "number": self.number,
            "title": self.title,
            "state": self.state,
            "url": self.url,
            "is_draft": self.is_draft,
            "head_ref_name": self.head_ref_name,
        })
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct RepoContext {
    pub git_branch: Option<String>,
    pub pr: Option<PullRequest>,
    pub availability: String,
    pub cwd: String,
}

impl RepoContext {
    pub fn to_detail_dict(&self) -> Value {
        json!({
            "cwd": self.cwd,
            "git_branch": self.git_branch,
            "pr": self.pr.as_ref().map(PullRequest::to_dict),
            "availability": self.availability,
        })
    }

    pub fn pr_summary(&self) -> Option<Value> {
        self.pr.as_ref().map(PullRequest::to_summary)
    }
}

#[derive(Clone)]
struct CacheEntry {
    context: RepoContext,
    expires_at: Instant,
}

#[derive(Clone)]
struct GhAuthState {
    ok: bool,
    checked_at: Instant,
}

type GhAuthCache = Mutex<Option<GhAuthState>>;

static CACHE: OnceLock<Mutex<HashMap<String, CacheEntry>>> = OnceLock::new();
static GH_AUTH: OnceLock<GhAuthCache> = OnceLock::new();

pub fn resolve_repo_context(cwd: &Path, refresh: bool) -> RepoContext {
    let original_cwd = cwd.to_string_lossy().to_string();
    let Ok(resolved) = cwd.canonicalize() else {
        return not_repo_context(original_cwd);
    };
    if !resolved.is_dir() {
        return not_repo_context(original_cwd);
    }
    let key = resolved.to_string_lossy().to_string();
    if !refresh {
        if let Some(entry) = cache().lock().unwrap().get(&key).cloned() {
            if entry.expires_at > Instant::now() {
                return entry.context;
            }
        }
    }

    let context = compute_context(&resolved);
    let ttl = if matches!(context.availability.as_str(), "ok" | "no-pr") {
        pr_ttl()
    } else {
        branch_ttl()
    };
    cache().lock().unwrap().insert(
        key,
        CacheEntry {
            context: context.clone(),
            expires_at: Instant::now() + ttl,
        },
    );
    context
}

pub fn current_git_branch(cwd: &Path) -> Option<String> {
    resolve_repo_context(cwd, false).git_branch
}

pub fn reset_caches_for_tests() {
    cache().lock().unwrap().clear();
    *gh_auth().lock().unwrap() = None;
}

fn compute_context(cwd: &Path) -> RepoContext {
    let (branch, is_repo) = resolve_branch(cwd);
    if !is_repo {
        return RepoContext {
            git_branch: None,
            pr: None,
            availability: "not-a-repo".to_string(),
            cwd: cwd.to_string_lossy().to_string(),
        };
    }
    let (pr, availability) = resolve_pr(cwd);
    RepoContext {
        git_branch: branch,
        pr,
        availability,
        cwd: cwd.to_string_lossy().to_string(),
    }
}

fn resolve_branch(cwd: &Path) -> (Option<String>, bool) {
    let (code, out, _) = run(
        &["git", "rev-parse", "--is-inside-work-tree"],
        cwd,
        branch_timeout(),
    );
    if code != 0 || !out.trim().eq_ignore_ascii_case("true") {
        return (None, false);
    }
    let (code, out, _) = run(
        &["git", "symbolic-ref", "--quiet", "--short", "HEAD"],
        cwd,
        branch_timeout(),
    );
    if code == 0 {
        let branch = out.trim();
        if !branch.is_empty() {
            return (Some(branch.to_string()), true);
        }
    }
    let (code, out, _) = run(
        &["git", "rev-parse", "--short", "HEAD"],
        cwd,
        branch_timeout(),
    );
    if code == 0 {
        let sha = out.trim();
        return ((!sha.is_empty()).then(|| sha.to_string()), true);
    }
    (None, true)
}

fn resolve_pr(cwd: &Path) -> (Option<PullRequest>, String) {
    if which("gh").is_none() || !check_gh_auth(false) {
        return (None, "no-gh".to_string());
    }
    let (code, out, err) = run(
        &[
            "gh",
            "pr",
            "view",
            "--json",
            "number,title,state,url,isDraft,headRefName",
        ],
        cwd,
        pr_timeout(),
    );
    if code == -1 {
        return (None, "error".to_string());
    }
    if code != 0 {
        let message = format!("{}{}", err, out).to_ascii_lowercase();
        if message.contains("no pull requests found")
            || message.contains("no pr")
            || message.contains("could not find")
        {
            return (None, "no-pr".to_string());
        }
        return (None, "error".to_string());
    }
    let Ok(data) = serde_json::from_str::<Value>(&out) else {
        return (None, "error".to_string());
    };
    let number = data
        .get("number")
        .and_then(Value::as_i64)
        .unwrap_or_default();
    if number <= 0 {
        return (None, "no-pr".to_string());
    }
    let pr = PullRequest {
        number,
        title: value_string(data.get("title")),
        state: value_string(data.get("state")).to_ascii_uppercase(),
        url: value_string(data.get("url")),
        is_draft: data
            .get("isDraft")
            .and_then(Value::as_bool)
            .unwrap_or(false),
        head_ref_name: value_string(data.get("headRefName")),
    };
    (Some(pr), "ok".to_string())
}

fn check_gh_auth(refresh: bool) -> bool {
    if which("gh").is_none() {
        *gh_auth().lock().unwrap() = Some(GhAuthState {
            ok: false,
            checked_at: Instant::now(),
        });
        return false;
    }
    if !refresh {
        if let Some(state) = gh_auth().lock().unwrap().clone() {
            if state.checked_at.elapsed() < gh_auth_ttl() {
                return state.ok;
            }
        }
    }
    let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    let (code, _, _) = run(&["gh", "auth", "status"], &cwd, pr_timeout());
    let ok = code == 0;
    *gh_auth().lock().unwrap() = Some(GhAuthState {
        ok,
        checked_at: Instant::now(),
    });
    ok
}

pub fn run_git_capture(
    cwd: &Path,
    args: &[&str],
    timeout: Duration,
    byte_cap: usize,
) -> (i32, String, String) {
    let mut command = Command::new("git");
    command.args(args).current_dir(cwd);
    run_command(command, timeout, byte_cap)
}

fn run(args: &[&str], cwd: &Path, timeout: Duration) -> (i32, String, String) {
    let mut command = Command::new(args[0]);
    command.args(&args[1..]).current_dir(cwd);
    run_command(command, timeout, 256 * 1024)
}

fn run_command(mut command: Command, _timeout: Duration, byte_cap: usize) -> (i32, String, String) {
    // std::process has no built-in timeout. Phase 2 keeps subprocess byte caps and
    // short command scopes; a later async process wrapper can enforce wall timeout.
    match command.output() {
        Ok(output) => (
            output.status.code().unwrap_or(1),
            capped_utf8(&output.stdout, byte_cap),
            capped_utf8(&output.stderr, byte_cap),
        ),
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
            (-1, String::new(), format!("missing executable: {err}"))
        }
        Err(err) => (-1, String::new(), err.to_string()),
    }
}

fn capped_utf8(raw: &[u8], cap: usize) -> String {
    let bytes = if raw.len() > cap { &raw[..cap] } else { raw };
    String::from_utf8_lossy(bytes).to_string()
}

fn not_repo_context(cwd: String) -> RepoContext {
    RepoContext {
        git_branch: None,
        pr: None,
        availability: "not-a-repo".to_string(),
        cwd,
    }
}

fn value_string(value: Option<&Value>) -> String {
    value
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

fn which(name: &str) -> Option<PathBuf> {
    let path = std::env::var_os("PATH")?;
    std::env::split_paths(&path)
        .map(|dir| dir.join(name))
        .find(|candidate| candidate.is_file())
}

fn env_duration(name: &str, default: f64) -> Duration {
    let seconds = std::env::var(name)
        .ok()
        .and_then(|value| value.parse::<f64>().ok())
        .filter(|value| value.is_finite() && *value >= 0.0)
        .unwrap_or(default);
    Duration::from_secs_f64(seconds)
}

fn branch_timeout() -> Duration {
    env_duration("CODEX_WEB_BRANCH_TIMEOUT_S", 2.0)
}

fn pr_timeout() -> Duration {
    env_duration("CODEX_WEB_PR_TIMEOUT_S", 4.0)
}

fn branch_ttl() -> Duration {
    env_duration("CODEX_WEB_BRANCH_TTL_S", 30.0)
}

fn pr_ttl() -> Duration {
    env_duration("CODEX_WEB_PR_TTL_S", 120.0)
}

fn gh_auth_ttl() -> Duration {
    env_duration("CODEX_WEB_GH_AUTH_TTL_S", 300.0)
}

fn cache() -> &'static Mutex<HashMap<String, CacheEntry>> {
    CACHE.get_or_init(|| Mutex::new(HashMap::new()))
}

fn gh_auth() -> &'static GhAuthCache {
    GH_AUTH.get_or_init(|| Mutex::new(None))
}
